File: NeetCode/Sliding_Window/Best_Time_To_Buy_and_Sell_Stock/best__time__to__buy_and__sell__stock.py
from typing import Optional, List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        h = [float("-inf")] * n
        l = [float("inf")] * n

        h[0] = prices[0]
        l[-1] = prices[-1]

        for i in range(1, n):
            h[i] = max(prices[i], h[i - 1])
            l[-i - 1] = min(prices[-i - 1], l[-i])

        maxProfit = 0
        for j in range(n):
            curr = prices[j]
            currProfit = h[j]
            if currProfit - curr > 0:
                maxProfit = max(maxProfit, currProfit)

        return maxProfit


if __name__ == "__main__":
    sol = Solution()
    prices = [10, 1, 5, 6, 7, 1]
    expected = 6
    print(sol.maxProfit(prices))
    print("Running Solution...")


"""
        class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        lowest = float("inf")

        lowest = min(prices[0], lowest)
        mxP = 0
        for index in range(len(prices)):
            p = prices[index] - lowest
            if p > 0:
                mxP = max(mxP, p)
            lowest = min(prices[index], lowest)
        return mxP

        """

File: NeetCode/Sliding_Window/Best_Time_To_Buy_and_Sell_Stock/Best_Time_To_Buy_and_Sell_Stock.cpp
// Best_Time_To_Buy_and_Sell_Stock.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int maxProfit(vector<int> &prices)
    {
        int minSell = prices[0];
        int maxProfit{0}, curr{0};

        for (int num : prices)
        {
            curr = num - minSell;
            maxProfit = max(maxProfit, curr);
            minSell = min(minSell, num);
        }
        return maxProfit;
    }
};
